packages/perseus/src/util/parse-perseus-json/general-purpose-parsers/discriminated-union.test.ts

```diff
@@ -0,0 +1,80 @@
+import {discriminatedUnion} from "./discriminated-union";
+import {constant} from "./constant";
+import {object} from "./object";
+import {number} from "./number";
+import {parse} from "../parse";
+import {failure, success} from "../result";
+
+describe("a discriminatedUnion with one variant", () => {
+    const unionParser = discriminatedUnion(
+        object({type: constant("ok")}),
+        object({type: constant("ok"), value: number}),
```

Comment:
This confused me until I went and looked at the implementation and docs. I guess without having a "shape" that represents the discriminant, we'd have to rebuild how TypeScript decides what it is.

Reply:
Yeah, I agree that this is ugly. For comparison, here's how Zod does it:

```ts
const myUnion = z.discriminatedUnion("status", [
  z.object({ status: z.literal("success"), data: z.string() }),
  z.object({ status: z.literal("failed"), error: z.instanceof(Error) }),
]);
```

I suppose we could do something like:

```ts
function discriminatedUnion<
  DiscriminantKey extends keyof any,
  T extends {[d: DiscriminantKey]: string | number | boolean}
>(discriminantKey: DiscriminantKey, parseVariant: Parser<T>): DiscriminatedUnionBuilder {
  // ...
}
```

That would mirror how TypeScript thinks about discriminated unions. But the problem is, it wouldn't handle our versioned widget options. Those aren't (in TS's view) valid discriminated unions, since the discriminant, `version.major`, is a nested field. My current parser design allows us to treat them kind of like discriminated unions anyway, by splitting the choice of variant and the parsing of that variant into separate steps that can run arbitrary parsing logic.

If only `version` were simply a _number_ (like it should have been all along), we could just do the thing that makes intuitive sense here. The curse of pre-TS code strikes again! :/

...and now, after writing all that out, I am tempted to create a special parser abstraction just for versioned widget options. Then we could simplify the one for discriminated unions to be more like Zod's API.

Reply:
Something like:

```ts
const parseMyWidget = parseVersionedWidgetOptions(parseVersion2)
  .migratedFrom(parseVersion1, upgradeFromVersion1)
  .migratedFrom(parseVersion0, upgradeFromVersion0)
  .parser
```